crates/runtime/src/model/embed.rs

```diff
@@ -13,20 +13,164 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+async fn get_bytes_for_file(
+    url: &str,
+    params: &HashMap<String, Secret<String>>,
+) -> Result<Bytes, Box<dyn std::error::Error + Send + Sync>> {
+    match url.split('/').collect_vec().as_slice() {
+        ["https:", "", "huggingface.co", org_id, model_id, "blob", branch, file @ ..] => {
+            get_file_from_hf(
+                None,
+                org_id,
+                model_id,
+                Some(branch),
+                file.join("/").as_str(),
+                params.get("hf_token").map(|s| s.expose_secret().as_str()),
+            )
+            .await
+        }
+        ["hf:", "", "datasets", org_id, model_id_revision, file @ ..] => {
+            let (model_id, branch) = parse_model_id_w_revision(model_id_revision);
+
+            get_file_from_hf(
+                Some("datasets"),
+                org_id,
+                model_id,
+                branch,
+                file.join("/").as_str(),
+                params.get("hf_token").map(|s| s.expose_secret().as_str()),
+            )
+            .await
+        }
+        ["hf:", "", "spaces", org_id, model_id_revision, file @ ..] => {
+            let (model_id, branch) = parse_model_id_w_revision(model_id_revision);
+            get_file_from_hf(
+                Some("spaces"),
+                org_id,
+                model_id,
+                branch,
+                file.join("/").as_str(),
+                params.get("hf_token").map(|s| s.expose_secret().as_str()),
+            )
+            .await
+        }
+        ["hf:", "", "models", org_id, model_id_revision, file @ ..] => {
+            let (model_id, branch) = parse_model_id_w_revision(model_id_revision);
+            get_file_from_hf(
+                Some("models"),
+                org_id,
+                model_id,
+                branch,
+                file.join("/").as_str(),
+                params.get("hf_token").map(|s| s.expose_secret().as_str()),
+            )
+            .await
+        }
+        ["hf:", "", org_id, model_id_revision, file @ ..] => {
+            let (model_id, branch) = parse_model_id_w_revision(model_id_revision);
+            get_file_from_hf(
+                Some("models"),
+                org_id,
+                model_id,
+                branch,
+                file.join("/").as_str(),
+                params.get("hf_token").map(|s| s.expose_secret().as_str()),
+            )
+            .await
+        }
+        _ => {
+            // Need to add `file://` for file paths
+            let final_url = match PathBuf::from_str(url).map(|p| p.canonicalize()) {
+                Ok(Ok(ref p)) if p.exists() => {
+                    format!("file://{}", p.to_string_lossy())
+                }
+                _ => url.to_string(),
+            };
+            let url = Url::parse(final_url.as_str()).boxed()?;
+            let (store, path) = object_store::parse_url(&url).boxed()?;
```

Comment:
We could use a custom error here to indicate invalid file path and provide user an actionable feedback